KEH-528 - Ability to select none

sebtheo opened the pull request:
This ticket sort of turned into Ability to select no**thing**~~ne~~.

------
Changelog:

- Added another button to the `survey` page that allows the user to go straight to the `validate_details` page. The pre-existing button will still redirect them to fill in any missing data and if all is filled in, the second button will be removed.
- `view_project` now renders each row with JS when doc loads. I couldn't find a way in Nunjucks to do this with the design system so went the JS route.
- Minimum needed to create a project is either the **Technical Contact** or **Delivery Manager** AND the **project name**
- Small fix for the icons on the `pre-survey` page

sebtheo committed:
button text unhidden

sebtheo committed:
add second button on /survey

sebtheo committed:
clean formScripts logic

sebtheo committed:
fallback when trying to retrieve data

sebtheo committed:
project.stage fix

sebtheo committed:
remove certain check logic

delivery manager or technical contact and project name are needed

sebtheo committed:
align icons in accordion

sebtheo committed:
minor changes to /survey logic

sebtheo committed:
view_project is dynamically rendered with JS

select none means missing data may occur and nunjucks doesnt have a good workaround for that

sebtheo committed:
rm print

sebtheo commented on application.py:
```diff
@@ -331,24 +331,29 @@
-    projects = projects.json()
+    try:
+        projects = projects.json()
+        print(projects)
```
Removed.

smstone0 commented on templates/survey.html:
```diff
@@ -106,12 +106,21 @@ <h1 class="ons-u-mt-m">Choose a section to complete</h1>
             "text": '<div id="submit-btn-text">Continue to Submission</div>',
             "classes": 'ons-u-mb-m ons-u-mt-l',
             "url": url_for('validate_details')
         }) }}
```
Unclear that this is the fill in missing details button, so could do with renaming the id and innerHTML as opposed to setting in the js.

sebtheo replied:
This was previous code. It only get's changed to 'Fill in missing details' if there are missing details, otherwise it remains as 'Continue to submission'. 

The other button is removed if there is no more data to be filled in.

You do raise a good point though that more explanation could be added to this.

smstone0 replied:
<img width="1670" alt="Screenshot 2025-03-05 at 13 59 19" src="https://github.com/user-attachments/assets/723293ae-9df1-460a-9bf5-ddbb734844be" />

I've given it a test and both buttons still appear even if all sections are marked as completed.

How would you feel about making each button single-purpose?

sebtheo replied:
Interesting, I don't get that? Each button is single purpose. 'Fill in missing details' goes to whatever isn't filled in, then the 'Continue to Submission' goes to the `/validate_details` page.

The secondary button (grey) is removed when the primary button (green) becomes 'Continue to Submission'.

![Screenshot 2025-03-05 at 14 11 52](https://github.com/user-attachments/assets/73da5751-30d8-4bdc-b0c5-02c67fdc9ecf)

smstone0 replied:
I think I found the issue. If you enter a language as primary (technology details), it will flag as completed for the section but the secondary button does not disappear. The issue goes away if you also enter a secondary language, which I'm guessing is what you did.

sebtheo replied:
Ah yes, I see that. I will fix, thank you.

sebtheo replied:
Fixed, thank you.

smstone0 replied:
Nice.

The buttons made a lot more sense when you called them primary and secondary, would you be happy for those ids to be renamed?

smstone0 replied:
<img width="708" alt="Screenshot 2025-03-05 at 16 03 56" src="https://github.com/user-attachments/assets/441d629d-3ad2-4586-9082-2ec2145392b9" />
'project_name' message could be made more user-friendly

sebtheo replied:
> Nice.
> 
> The buttons made a lot more sense when you called them primary and secondary, would you be happy for those ids to be renamed?

Renamed [here](https://github.com/ONS-Innovation/keh-tech-audit-tool/pull/14/commits/d5f102acd6c24400e550af22f1a7d9e208e3bbf4).

sebtheo replied:
> <img alt="Screenshot 2025-03-05 at 16 03 56" width="708" src="https://private-user-images.githubusercontent.com/106412444/419534238-441d629d-3ad2-4586-9082-2ec2145392b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExOTQ3OTksIm5iZiI6MTc0MTE5NDQ5OSwicGF0aCI6Ii8xMDY0MTI0NDQvNDE5NTM0MjM4LTQ0MWQ2MjlkLTNhZDItNDU4Ni05MDgyLTJlYzIxNDUzOTJiOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwNVQxNzA4MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03OThkMmY3ZGRjYmVkYjVkNGRiZjYxZDNhNDE3MTM0OWFmM2Y2ZDZhYjAzZTk1NTY2YzcyY2MxOWVhMmNjNzJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cAMXb3Q3ia8uY82CAGY-pXAHww-6JvSIBxuj0vNQLPY"> 'project_name' message could be made more user-friendly

Updated [here](https://github.com/ONS-Innovation/keh-tech-audit-tool/pull/14/commits/578dc4a7e3db29a52ad87e3d3ea2db243c6a5682). This is *kind of* a quick fix. Although we this is the only required field so the only possible `field` that is passed is `project_name`. If these requirements change (more required fields) then we can adjust then. This works though.

smstone0 replied:
Looks good 👍🏻

sebtheo committed:
update validation check for languages

sebtheo committed:
display more user friendly error

proejct_name is the only required field that could be flagged (delivery/technical contact are already accounted for) so this fix works with the current system

any other required fields would have to be accounted for. previously we have mapped this using a dict, but there is no need for it currently.

sebtheo committed:
rename primary and secondary btns

delterr commented:
(1) When there is a missing piece of data on the Technology Details section and you get to the summary page, it should have a button to continue anyways, and shouldnt have an error pop up
<img width="1792" alt="image" src="https://github.com/user-attachments/assets/0440b09b-1709-423e-ae33-ea331be838f0" />

(2) When there is a project with incomplete data, it seems to break rendering for the edit page. 
<img width="1784" alt="image" src="https://github.com/user-attachments/assets/627889d7-cd42-4b91-82ed-74797008d158" />

Looking at the console, it seems to have trouble parsing out the original_data 
<img width="522" alt="image" src="https://github.com/user-attachments/assets/d883a965-2a4f-4ade-9d53-bb259f92238e" />

delterr reviewed (changes requested):
(1) On the technology section, error messages show when not needed and there is no continue button

(2) Rendering on Edit page breaks when there is missing data

sebtheo committed:
remove validation in summary pages

sebtheo committed:
remove repeated height styling code

sebtheo committed:
fix None error when loading edit page

delterr commented:
Also not sure if you're already doing this, but please use the `-testing` bucket when testing locally, I want to keep dev and local as separate as possible.

sebtheo committed:
reformatting certain areas

sebtheo committed:
fix post py code

sebtheo committed:
fix formatting and loading errors

sebtheo committed:
rm rogue print

delterr reviewed (changes requested):
There seems to be this weird bug where if you were to edit something twice (you dont even need to change anything), the technical contact and delivery manager won't be picked up by validate_details on the edit page. Also another weird thing that happens is that the doc link gets wrapped in quotes and square brackets each time you save from the edit page

<img width="295" alt="image" src="https://github.com/user-attachments/assets/064587f5-57c0-4831-8931-4ff9d7dc8fd9" />

sebtheo committed:
fix render data and save data

sebtheo committed:
Merge branch 'main' into select-none

sebtheo committed:
rm print

sebtheo committed:
render project description on validate details

sebtheo committed:
validation on render source control links

sebtheo committed:
fallback for doc link submission

gibbardsteve commented on static/formScripts.js:
```diff
@@ -113,239 +113,170 @@ function updateSectionStatus(sectionId, dataItems) { // Update selection status
+    }
+    
+    // Define all validation checks with their URLs and validation functions
+    const validations = [
```
I like the way you have done this.

sebtheo replied:
Thank you

gibbardsteve commented on templates/chapter_summaries/architecture_summary.html:
```diff
@@ -113,59 +97,41 @@ <h1 class="ons-u-mt-m">Code and Architecture Summary</h1>
                         try {
```
It looks like we can remove this? I don't see any operations now you have removed the errorPanel stuff

gibbardsteve commented on templates/chapter_summaries/architecture_summary.html:
```diff
@@ -113,59 +97,41 @@ <h1 class="ons-u-mt-m">Code and Architecture Summary</h1>
                             document.getElementById('source_control_details').innerHTML = source_control;
                         } catch (e) {
-                            errorList.push('Source Control');
+                            // No error
```
We either want to catch and error or we don't. if JSON.parse fails for some reason we should at least log an error I think.

gibbardsteve commented on templates/chapter_summaries/architecture_summary.html:
```diff
@@ -113,59 +97,41 @@ <h1 class="ons-u-mt-m">Code and Architecture Summary</h1>
                             document.getElementById('database_details').innerHTML = database_details;
                         } catch (e) {
-                            errorList.push('Database');
+                            // No error
```
As above

gibbardsteve commented on templates/chapter_summaries/architecture_summary.html:
```diff
@@ -113,59 +97,41 @@ <h1 class="ons-u-mt-m">Code and Architecture Summary</h1>
                             document.getElementById('hosting_details').innerHTML = hosting_details;
                         } catch (e) {
-                            errorList.push('Hosting');
+                            // No error
```
As above

gibbardsteve commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,119 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
+                                }
+                                document.getElementById('developed_details').innerHTML = developed_details;
+                                document.getElementById('developed').value = JSON.stringify(developedDataJSON.developed);
+                            } catch (e) { // No error
```
Empty catch

gibbardsteve commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,119 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
+                                let stageDetailsJSON = JSON.parse(stageData).stage;
+                                document.getElementById('stage_details').innerHTML = stageDetailsJSON;
+                                document.getElementById('stage').value = JSON.stringify(stageDetailsJSON);
+                            } catch (e) { // No error
```
empty catch

gibbardsteve commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,119 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
+                                }
+                                document.getElementById('user_details').innerHTML = user_details;
+                                document.getElementById('user').value = JSON.stringify(techContactDataJSON);
+                            } catch (e) { // No error
```
empty catch

gibbardsteve commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,119 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
+                                }
+                                document.getElementById('project_details').innerHTML = project_details;
+                                document.getElementById('project').value = JSON.stringify(projectDataJSON);
+                            } catch (e) { // No error
```
empty catch

gibbardsteve commented on templates/chapter_summaries/supporting_tools_summary.html:
empty catch legs. Can we log something?

gibbardsteve commented on templates/chapter_summaries/tech_summary.html:
```diff
@@ -124,12 +109,10 @@ <h1 class="ons-u-mt-m">Technology Details Summary</h1>
                         var languagesData = localStorage.getItem('languages-data');
-                        var errorList = [];
                         localStorage.setItem('validate', false);
                         try {
```
Remove?

gibbardsteve commented on templates/chapter_summaries/tech_summary.html:
```diff
@@ -168,59 +143,51 @@ <h1 class="ons-u-mt-m">Technology Details Summary</h1>
                                 languages_details += `<br>Other: ${languagesDataJSON.others.map(item => `${item}`).join(', ')}`;
                             }
                             if (languagesDataJSON.main.length === 0 && languagesDataJSON.others.length === 0) {
```
Remove?

gibbardsteve commented on templates/chapter_summaries/tech_summary.html:
```diff
@@ -168,59 +143,51 @@ <h1 class="ons-u-mt-m">Technology Details Summary</h1>
 
                         try {
                             var frameworksDataJSON = JSON.parse(frameworksData);
                             if (frameworksDataJSON.others.length === 0) {
```
Remove

gibbardsteve commented on templates/chapter_summaries/tech_summary.html:
empty catch legs

gibbardsteve commented on templates/validate_details.html:
```diff
@@ -486,465 +506,799 @@ <h1 class="ons-u-mt-m">Check your answers</h1>
+        /**
+         * Core utility functions for data processing
+         */
+        const DataUtils = {
```
I don't want you to change anything (we all agree the tech audit tool may be overhauled at some point) but it feels like when we are defining swathes 'utlities' in javascript (which includes functionality to map to from api format etc) inside a html file, it would be better to create a scripts.js and have these functions defined in there?  Again, I don't want us to change anything now - I am just checking that as a pattern that seems reasonable.

sebtheo replied:
Yeah, we have formScripts that is used in the pages where data entry is used. The validate details only uses this JS code. We should move it out technically.

gibbardsteve commented on templates/view_project.html:
```diff
@@ -63,265 +49,276 @@
                                   "rowItems": [
                                       {
-                                          "id": "sales-dates",
+                                          "id": "technical-contact",
```
It is a good catch that we have changed these. At some point Hasan had some selenium tests running, have you rn these with your updated code?

delterr replied:
I haven't tested these changes but i'm pretty sure those id's were irrelevant and not being called by any Selenium tests or `document.getElementByID`. Still worth running the tests one more time to make sure.

sebtheo replied:
Updated the testing. The ID wasn't an issue but programme short/name needed adding and general linting.

gibbardsteve reviewed (commented):
Is this code on dev? I have only looked at the source.
A couple of minors about empty catch legs and a query as to whether we have re-run Hasan's selenium tests?

sebtheo committed:
log errors

sebtheo commented:
I've added back in the logging. Basically, if a field at the beginning wasn't filled in then the rest of them wouldn't fill in. The empty catches aren't really highlighting that there is an error, it's more just making sure the rest can fill in even if a field before hasn't been filled in. I have added them back though.

sebtheo committed:
add escapeHTML to project summary

sebtheo committed:
lint application.py

sebtheo committed:
rm unused test

sebtheo committed:
add linting for tests folder

sebtheo committed:
lint tests and update create project

sebtheo commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,123 @@
+                            }
+                            try {
+                                let stageDetailsJSON = JSON.parse(stageData).stage;
+                                document.getElementById('stage_details').innerHTML = stageDetailsJSON;
```
Fixed this

sebtheo committed:
sanitise project_name

TotalDwarf03 commented on application.py:
```diff
@@ -419,58 +435,85 @@ def survey():
-    elif form_data["developed"]["developed"] == "Partnership":
-        developed_company = form_data["developed"]["partnership_company"]
-
+    if form_data.get("developed") != {}:
```
form_data.get("developed") is repeated.

Put it in a variable and reference that.

developed = form_data.get("developed", "")

TotalDwarf03 commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,123 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
+                                        .display = 'none';
+                                }
+                                if (projectDataJSON.short_name !== "") {
+                                    project_details += `Project Short Name:<p style="font-weight: 400;">${
```
inline styling is repeated. This should probably be moved to a class and referenced there.

TotalDwarf03 commented on templates/chapter_summaries/project_summary.html:
```diff
@@ -122,122 +122,123 @@ <h1 class="ons-u-mt-m">Project Details Summary</h1>
-                                .getElementById('submit-button')
-                                .style
-                                .display = 'none';
+                    <script>
```
Echoing that this JS logic should be in a separate `.js` file

TotalDwarf03 commented on templates/section_project/developed.html:
```diff
@@ -120,30 +117,28 @@ <h1>Developed</h1>
+                    if (Array.isArray(data)) {
+                        option = obj[data[0]];
+                        if (data[0] === "Outsourced") {
+                            document.getElementById('other-input-1').value = data[1] || "";
```
Is `other-input-1` a sensible id?

TotalDwarf03 commented on templates/survey.html:
```diff
@@ -106,12 +106,21 @@ <h1 class="ons-u-mt-m">Choose a section to complete</h1>
     {{ onsButton({
-            "id": "submit-btn",
-            "text": '<div id="submit-btn-text">Continue to Submission</div>',
+            "id": "primary-btn",
```
should `primary-btn` and `secondary-btn` be a bit more descriptive?

primary-submit-btn perhaps?

TotalDwarf03 commented on templates/validate_details.html:
```diff
@@ -486,465 +506,799 @@ <h1 class="ons-u-mt-m">Check your answers</h1>
+                    details += 'Project Name: <p style="font-weight:400">' + name + '</p>';
                 }
+                if (shortName) {
+                    details += 'Short Name: <p style="font-weight:400">' + shortName + '</p>';
```
Ditto on repeated inline styling. Move this to a class if you can.

TotalDwarf03 commented on templates/validate_details.html:
```diff
@@ -486,465 +506,799 @@ <h1 class="ons-u-mt-m">Check your answers</h1>
+        /**
+         * Data normalization and conversion
+         */
+        const DataNormalizer = {
```
again, should probably look to extract the js into a separate file.

TotalDwarf03 reviewed (commented):
Couple of comments. Minor changes

sebtheo committed:
fix developed logic

sebtheo committed:
change developed input ids

sebtheo committed:
major reformat

Take JS out of validate_details and put in separate JS, same for summary pages.

Fix my misunderstanding of how developed was stored.

It is stored as a list of strings ["", ""]

list[0] is In-house/Outsourced/Partnership then the list[1] is either empty or the other company

sebtheo committed:
change button ids

sebtheo committed:
escapeHTML content

smstone0 commented:
Testing on dev - this is what I see on the summary pages
![{4095E11A-031B-4994-894A-A699E9CCB2E6}](https://github.com/user-attachments/assets/65323f95-0212-4f98-8f62-f9332db4148b)
![{1CA688FA-9B54-4798-BA39-8FE3FF160E80}](https://github.com/user-attachments/assets/0a311519-4d09-4ae4-b632-01a6f068d15f)

sebtheo committed:
fix displaying HTML

sebtheo commented:
> Testing on dev - this is what I see on the summary pages ![{4095E11A-031B-4994-894A-A699E9CCB2E6}](https://private-user-images.githubusercontent.com/106412444/422834761-65323f95-0212-4f98-8f62-f9332db4148b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMDE5NzEsIm5iZiI6MTc0MjIwMTY3MSwicGF0aCI6Ii8xMDY0MTI0NDQvNDIyODM0NzYxLTY1MzIzZjk1LTAyMTItNGY5OC04ZjYyLWY5MzMyZGI0MTQ4Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxN1QwODU0MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YmU5ZjMzZDc1MjRiNzliMDg0OWYxZDU1ZmY2YTQ0Mjg0NTM3N2NmNDFmZThmMDFlODAyNjk3NjhmZjY2Yzk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FVm-IWOOvCuNF3DKISbNKjQgMrfZhJb9a_xs3T-S-xk) ![{1CA688FA-9B54-4798-BA39-8FE3FF160E80}](https://private-user-images.githubusercontent.com/106412444/422834873-0a311519-4d09-4ae4-b632-01a6f068d15f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMDE5NzEsIm5iZiI6MTc0MjIwMTY3MSwicGF0aCI6Ii8xMDY0MTI0NDQvNDIyODM0ODczLTBhMzExNTE5LTRkMDktNGFlNC1iNjMyLTAxYTZmMDY4ZDE1Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxN1QwODU0MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjAxMTkyOThiOWQyMDBiMThlZmEzMzVmZTBkZWQzNjg2YWRmMjU2ZjljZmYyOWIxMjE5ZjI4YTlhZjZiMGIyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.LC-RAgIWDX0A9Ac83el3UOKKp9lmHPFIyfPpRmkl7Vc)

Oops, fixed it.

smstone0 commented:
![{6755C486-A678-4B43-AE3A-524E261289D6}](https://github.com/user-attachments/assets/d6dfb358-c392-42f7-bfc8-046ab16990b8)
~~Testing on dev again and the sub-headings e.g. above are missing, guessing this isn't expected behaviour?~~
Re-logged in and the sub-headings are back

![{4FAA8F23-B41A-4B43-8ECF-4D54BAA977DD}](https://github.com/user-attachments/assets/6a3d68ac-c1c8-4386-a26b-13453f394ec9)
There's also a bug where I can select a grade from the radio button but not provide an actual technical contact/delivery manager email address and it validates as having been provided. Not an issue on the main summary page though.

TotalDwarf03 commented on application.py:
```diff
@@ -51,43 +53,40 @@ def read_auto_complete_data():
         raise e
 
-    secret = secret_value_response.get('SecretString')
+    secret = secret_value_response.get("SecretString")
```
Should this do something if `.get()` returns `None`?

Are we ok with a key error being raised if the secret doesn't exist on line 89?

Maybe we could `flash()` an appropriate error message?

TotalDwarf03 commented on application.py:
```diff
@@ -313,50 +318,57 @@ def dashboard():
+            flash(f"Project not found. Please try again. {projects['message']}")
             return redirect(url_for("dashboard"))
-    except Exception:
+    except:
```
Can we change this except to only allow the specific error that happens within the above?

I believe it will be a `KeyError` if `projects["message"]` doesn't exist

TotalDwarf03 commented on static/js/submission.js:
```diff
@@ -0,0 +1,670 @@
+        let result = '';
+        if (data.developed) {
+            // Object format
+            result += 'Developed: <p style="font-weight:400">' + data.developed + '</p>';
```
repeated inline styling. Move this to a class please

TotalDwarf03 commented on static/js/summary.js:
```diff
@@ -0,0 +1,523 @@
+    static updateElement(elementId, content) {
+        const element = document.getElementById(elementId);
+        if (element) {
+            element.innerHTML = content;
```
Are we planning to implement this or are we calling `SummaryUtils.escapeHtml()` before passing into the function?

Even if we are, it might be worth doing this fix as a just in case.

TotalDwarf03 commented on templates/section_project/developed.html:
```diff
@@ -79,75 +79,120 @@ <h1>Developed</h1>
-                    data.partnership_company = document.getElementById('other-input-2').value;
-                } else {
-                    data.complete = false;
+            // Call redirectToPrevious to set up the return path
```
This should also probably be in its own js file.

TotalDwarf03 commented on templates/section_project/developed.html:
```diff
@@ -79,75 +79,120 @@ <h1>Developed</h1>
+                            }
+                        }
+                    } else if (data.developed) {
+                        // Legacy object format
```
Do we need to maintain a legacy format when our dataset is very small at the moment?

Could we convert the old format to the new one while our dataset is small?

If we start to have legacy support stuff at this early stage, out data will get out of control very quickly. Standardise before the dataset grows.

TotalDwarf03 commented on templates/view_project.html:
```diff
@@ -336,4 +333,158 @@
\ No newline at end of file
+{% endblock %}
+{% block scripts %}
+<script>
```
Can we move this to a js file please?

TotalDwarf03 reviewed (commented):
Couple of comments. 

I noticed that the js functions don't really have any docstrings and most functions are unlabelled besides the function name. I haven't picked these up in the review as we'd end up with lots of comments but can we keep this in mind when doing js stuff here in future? 

Docstrings are important to provide context for a function.

sebtheo committed:
fix TL and DM display on project summary

sebtheo commented:
> ![{6755C486-A678-4B43-AE3A-524E261289D6}](https://private-user-images.githubusercontent.com/106412444/423536370-d6dfb358-c392-42f7-bfc8-046ab16990b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyOTA4NTksIm5iZiI6MTc0MjI5MDU1OSwicGF0aCI6Ii8xMDY0MTI0NDQvNDIzNTM2MzcwLWQ2ZGZiMzU4LWMzOTItNDJmNy1iZmM4LTA0NmFiMTY5OTBiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQwOTM1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjU1MjExYTNjNTM1YTllNTY5ZjI0NDA3MmQ3MzhhMTg5OTVhYzgzNTllMjI2YjIxNmQyN2ZkNWRkNTUxZmMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9rYdSJ2FFuHFc6H4M42yEFoknt1n5gsa4ocBegs5RxY) ~Testing on dev again and the sub-headings e.g. above are missing, guessing this isn't expected behaviour?~ Re-logged in and the sub-headings are back
> 
> ![{4FAA8F23-B41A-4B43-8ECF-4D54BAA977DD}](https://private-user-images.githubusercontent.com/106412444/423538121-6a3d68ac-c1c8-4386-a26b-13453f394ec9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyOTA4NTksIm5iZiI6MTc0MjI5MDU1OSwicGF0aCI6Ii8xMDY0MTI0NDQvNDIzNTM4MTIxLTZhM2Q2OGFjLWMxYzgtNDM4Ni1hMjZiLTEzNDUzZjM5NGVjOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQwOTM1NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNWExYWIzMTNlNjk2YjBiZDljMGU4Y2Q0ZDEwMDVkMzYyNGYzY2MzMmRlM2M3ZTFlOThlZTViN2JhZWE0MTdhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.f-DGe79-SgJjwgb1bhiZ4TCbZedOQwZHjPgCqZryAoE) There's also a bug where I can select a grade from the radio button but not provide an actual technical contact/delivery manager email address and it validates as having been provided. Not an issue on the main summary page though.

Thank you. I've added a fix :)

sebtheo committed:
fallback when getting SecretString

sebtheo committed:
more detailed exception

sebtheo committed:
function to build p element

sebtheo committed:
view_project js to external file

sebtheo committed:
developed to external js file

sebtheo committed:
rm legacy format for developed obj

TotalDwarf03 reviewed (approved):
👍

sebtheo committed:
Update test_project_creation.py

https://random-word-api.herokuapp.com/word returns:

java.lang.NoClassDefFoundError: Could not initialize class play.api.libs.json.Writes$

so I switched the random word API provider

sebtheo committed:
small cleanup

sebtheo committed:
testing choice GCP

sebtheo committed:
Merge branch 'main' into select-none

sebtheo committed:
project_summary fix

sebtheo committed:
developed showing when empty
